### PR TITLE
Various RTL improvements (mainly VTOL specific)

### PR DESCRIPTION
**Pull request by sfuhrer**
Replaces https://github.com/PX4/PX4-Autopilot/pull/14728.

**Describe problem solved by this pull request**
- currently landing on the landing pattern doesn't seem to work (land pattern = whatever of the mission is after the DO_LAND_START marker)
- high transition altitude (see https://github.com/PX4/PX4-Autopilot/pull/14728 for context)

**Describe your solution**
RTL logic with this PR, changes to current logic in **bold**
(assuming RTL_TYPE is set to 1 (VTOL and FW default, using mission landing if available))
- if close to RTL landing position and in hover, use cone RTL logic (if mission landing is planned that one is used)
- if VTOL and in hover mode, use normal MC RTL strategy (climb to `RTL_RETURN_ALT`, come back and land at RTL landing position)
- if DO_LAND_START is set (landing using mission landing): 
  - if in FW mode, go to DO_LAND_START marker in FW mode and from there on follow mission landing
  - if in MC mode, go to the Land position and land there (don't follow landing pattern as in FW mode)
-  if DO_LAND_START is not set 
   - if VTOL and in FW mode, climb to `RTL_RETURN_ALT`, fly to home, **loiter down until `RTL_DESCEND_ALT` in FW mode**, transition to hover and land

**Test data / coverage**
SITL tested with Standard VTOL and multicopter. 

**Additional context**
For VTOL, it is recommended to always fly with a mission landing planned, such that this one can be used during the RTL. Landing approaches using mission landing can be optimized for current wind direction (transition into the wind), and result in less hover time (vehicle transitions over hover, instead of while loitering around home)

### CODE REVIEW

**Comment by Antiheavy** [comment]
is this:


> For VTOL, it is recommended to always fly with a mission landing planned, such that this one can be used during the RTL. Landing approaches using mission landing can be optimized for current wind direction (transition into the wind), and result in less hover time (vehicle transitions over hover, instead of while loitering around home)

consistent with this?:


> * if close to home position and in hover, use cone RTL logic
> 
> * if VTOL and in hover mode, use normal MC RTL strategy (climb to `RTL_RETURN_ALT`, come back and land)

Have you considered if there should be a parameter that allows RTL at the Home location if in MC mode?  I wonder how many VTOL people would always want to use the mission landing if at all possible?  The "return to home" fall-back in the RTL_TYPE=1 logic was originally designed for the case when the mission landing accidentally gets deleted or gets overwritten with an infeasible landing pattern.  

From the way this PR looks to be currently written you intend the VTOL vehicle will perform different Return mode behaviors depending on the context.  I wonder if that is how 100% of VTOL users will want it to work.  I'm not a VTOL expert so I'm not advocating one way or another, just curious.

**Comment by sfuhrer** [comment]
> Have you considered if there should be a parameter that allows/avoids RTL at the Home location if in MC mode?

Thanks for pointing this out. I actually wrote it down wrongly in the description. It will **always use the mission landing for RTL if RTL_TYPE is 1**. What changes is that it now won't follow the landing pattern (all that is after the DO_LAND_START) in hover, but directly go to Land and descend there.

File: src/modules/navigator/mission.cpp
Comment (dagar) [inline src/modules/navigator/mission.cpp:81]: I'm not sure I follow, how would this fail if the mission landing progress was set to false immediately on first inactivation?

File: src/modules/navigator/rtl.cpp
Comment (dagar) [inline src/modules/navigator/rtl.cpp:323]:
```suggestion
			// even if current climb altitude is below (e.g. RTL immediately after take off)
```

File: src/modules/navigator/rtl_params.c
Comment (dagar) [inline src/modules/navigator/rtl_params.c:137]: Now I'm wondering why this wasn't already set by default.

Comment (sfuhrer) [reply src/modules/navigator/rtl_params.c:137]: Yes that's a big wonder. I found the cone RTL always really useful so far, both for VTOL in hover and MC.

File: src/modules/navigator/mission.cpp
Comment (RomanBapst) [reply src/modules/navigator/mission.cpp:81]: @dagar This prevents the situation where e.g. you are on a mission landing and then RTL kicks in. You don't want to start from the beginning but continue the mission. However, if e.g. the user takes over in manual during a mission landing and then switches back, you probably want to start he mission landing from scratch, as you have no clue where exactly you are and if you are still aligned with the path.

**Review by RomanBapst — APPROVED** [review]

File: src/modules/navigator/mission.cpp
Comment (dagar) [reply src/modules/navigator/mission.cpp:81]: I should have been more specific, the part I don't follow is the magic 2 seconds.

Comment (RomanBapst) [reply src/modules/navigator/mission.cpp:81]: @dagar A long study over many years has calculated this number to be the one and only reasonable value. It's almost like pi....

Comment (dagar) [reply src/modules/navigator/mission.cpp:81]: Why not do it immediately on the actual mission deactivation?

Comment (dagar) [reply src/modules/navigator/mission.cpp:81]: This is for mission -> RTL (failsafe, etc) -> mission (planned landing)?

**Comment by moreba1** [comment]
in master version, VTOL has a dangerous behavior when it reaches VTOL Land.
@sfuhrer 
https://logs.px4.io/plot_app?log=9df156fc-233e-4b54-ad23-e82017ff47f4

**Comment by dagar** [comment]
Rebased on master.

**Comment by yuthikasagarage** [comment]
I also noticed that the standard_vtol back transition is behaving dangerously. it's on master and also in this pr,
I noticed this on my flight on the real hardware as well. 
just after the back-transition it tries to do a roll maneuver in most of the times,
But every once in a while, it does a ok back-transition, without trying to roll . 

same mission, two instances, same-pr (firmware), back-transition behavior is different.


https://user-images.githubusercontent.com/25317204/103393299-f0d23000-4b47-11eb-8937-cd642d604219.mp4

https://review.px4.io/plot_app?log=1e1bb5d2-5260-40da-bd7e-46fadf55fc6c ---> rolling
https://review.px4.io/plot_app?log=12468162-8610-4965-831c-dd25c8edfeda ---> without roll

**Comment by LorenzMeier** [comment]
Thanks for reporting - @RomanBapst @sfuhrer could you please follow up?

**Comment by LorenzMeier** [comment]
@yuthikasagarage Have you reviewed your flights? It looks like the vehicle is overshooting the transition waypoint (so your plan is "not doable") and hence correcting. There is nothing wrong about the control performance, this seems to be more a planning UX / tuning issue than anything else. That still needs to be solved, but not necessarily in the VTOL controller.

**Comment by yuthikasagarage** [comment]
i will look into the flights again, and will try to tune the back transition and check whether the issue persists.
however, with a bad planning let's say it does a back-transition in some case, i think all the back-transitions need to be on level flight even if it overshoots, just my two cents.

File: src/modules/navigator/mission.cpp
Comment (LorenzMeier) [reply src/modules/navigator/mission.cpp:81]: Let's track this in a separate issue - we shouldn't be blocking an incremental move towards a much better behavior.

**Review by LorenzMeier — APPROVED** [review]
Reviewed, approving after intense flight testing.

**Comment by sfuhrer** [comment]
> in the master version, VTOL has a dangerous behavior when it reaches VTOL Land.

This should be fixed with this PR, please re-test.

> desirable behavior is that VTOL transition location be near to land location to minimize power consumption, and the drone doesn't have that much tilt angle and oscillation in MC mode.

With the code from this PR in, it will transition at a distance of the loiter radius from home if no mission landing is planned and the waypoint before the landing waypoint isn't at the transition altitude (RTL_DESCEND_ALT). I'm aware that this isn't optimal for minimized hover time, but at least it should be an improvement to before where it could transition 100m above home and then descend in hover.

File: src/modules/navigator/mission.cpp
Comment (sfuhrer) [reply src/modules/navigator/mission.cpp:81]: Issue to track that: https://github.com/PX4/PX4-Autopilot/issues/16488

**Comment by moreba1** [comment]
>     * if in FW mode, go to DO_LAND_START marker in FW mode and from there on follow mission landing

Is it possible that VTOL first climbs to RTL_RETURN_ALT and keep this attitude until it reaches to landing pattern?

**Comment by sfuhrer** [comment]
>Is it possible that VTOL first climbs to RTL_RETURN_ALT and keep this attitude until it reaches to landing pattern?

That should be the desired behavior. If you noticed something else then let me know, that would indicate a bug.

File: src/modules/navigator/mission.cpp
Comment (RomanBapst) [reply src/modules/navigator/mission.cpp:81]: @dagar I think I might need to explain this a bit better:
When you are executing a mission landing and RTL kicks in due to low battery, then the mode switches from mission to RTL and back to mission (assuming you are executing a mission landing, which is the preferred choice almost always).
Two the second delay (it actually needs way less) allows the mission to just continue the landing rather than restarting the landing pattern.

Comment (RomanBapst) [reply src/modules/navigator/mission.cpp:81]: @dagar Sorry, I noticed I told you this before. What is the issue here?

Comment (dagar) [reply src/modules/navigator/mission.cpp:81]: > @dagar Sorry, I noticed I told you this before. What is the issue here?

Nothing necessarily, I just wanted to make sure there wasn't a potential hole for a hard to reproduce bug down the round. Eg dataman chokes reading the sdcard and there's a one off delay that's much longer than expected. 
￼

**Comment by VTOLDavid** [comment]
> 
> 
> i will look into the flights again, and will try to tune the back transition and check whether the issue persists.
> however, with a bad planning let's say it does a back-transition in some case, i think all the back-transitions need to be on level flight even if it overshoots, just my two cents.

I having this problem with a tailsitter with very dangerous turns during transition. Is it possible to disable the navigation during transition? The sugested solution is to give more time to the backtransition?

**Comment by sfuhrer** [comment]
> I having this problem with a tailsitter with very dangerous turns during transition. Is it possible to disable the navigation during transition? The sugested solution is to give more time to the backtransition?

I don't think this is directly related to this specific PR - would you mind creating a new issue, and include some flight logs in the description? Thanks!

**Comment by hamishwillee** [comment]
@sfuhrer @Antiheavy Are there any docs changes related to this for PX4v1.12? My guess is yes, though it is unclear with the long thread, what delivered/what the landing behaviour is (?).

Looking at the initial post, it is mostly about what should happen if RTL_TYPE is set to use a mission landing (RTL_TYPE=1) and there was no mission landing pattern defined - in which case it should fly down in FW mode to the descent altitude and then switch to MC and descend "as per usual". 

The description is a little flawed because it does not cater for the behaviour with rally points. 

Docs for vtol are here:
https://docs.px4.io/master/en/flight_modes/return.html#vtol
https://docs.px4.io/master/en/flight_modes/return.html#mission-landing-rally-point-return-type-rtl-type-1

**Pull request merged**
